server/src/main/java/org/apache/calcite/avatica/server/HttpServer.java

```diff
@@ -239,18 +239,8 @@ protected void internalStart() {
     server.setConnectors(new Connector[] { connector });
 
     // Default to using the handler that was passed in
-    final HandlerList handlerList = new HandlerList();
-    Handler avaticaHandler = handler;
-
-    // Wrap the provided handler for security if we made one
-    if (null != securityHandler) {
-      securityHandler.setHandler(handler);
-      avaticaHandler = securityHandler;
-    }
+    configureHandlers(securityHandler);
```

Comment:
See my other comment. While what we care about is customizing security handlers, since we have to think about generic interface. Is it enough just to say CUSTOM means that the user will supply a list of handlers including whatever security handler they want to use, so we don't build anything canned ourselves? We are pretty close to this already.

Reply:
I agree, in order to comply with `CUSTOM` as the authentication method, we need to also allow users to specify that. The builder allows users to add their handlers, but still wraps them around with the `securityHandler` which gets configured based on the auth type. 

A better of approaching this would be to bypass all these code for `CUSTOM` security config, so that it doesn't bring any expectations to the user. What do you think?

Reply:
> A better of approaching this would be to bypass all these code for CUSTOM security config, so that it doesn't bring any expectations to the user.

Yes, exactly

Reply:
I can go ahead and add that change here, @joshelser Please advice.

Reply:
I like your suggestion, @apurtell! I was a bit reticent to suggest that as, like you imply, it's a bit more work to set up all of those handlers on your own. However, this would also fit nicely into the work that @aaraujo has already done (the parameterized Jetty-specific stuff for the `Builder` class).

I think that providing "well-known authentication scenarios" is a nice thing that we can do (e.g. basic, digest, spengo), and then leave `CUSTOM` for the "let me drive" scenario where you set it all up yourself.

I'm still a bit curious if there's another "nicer" abstraction we could provide from an API perspective that isn't full CUSTOM. For example, are there more abstractions which would help us put the boilerplate into the HttpServer.Builder and let you inject your SFDC specific authn/authz decisions? I am just a little ignorant on the specifics of how your SFDC-specific thing works. You can also just tell me to drop it and I'll stop poking :)

Reply:
> I like your suggestion, @apurtell! I was a bit reticent to suggest that as, like you imply, it's a bit more work to set up all of those handlers on your own

Agreed and that's probably the reason why I forgot abstracting them out as well :)

> I'm still a bit curious if there's another "nicer" abstraction we could provide from an API perspective that isn't full CUSTOM. For example, are there more abstractions which would help us put the boilerplate into the HttpServer.Builder and let you inject your SFDC specific authn/authz decisions? I am just a little ignorant on the specifics of how your SFDC-specific thing works. You can also just tell me to drop it and I'll stop poking :)

@joshelser I initially wanted to do that, however `AvaticaServerConfiguration` is not abstracted out completely. I will file a Jira to refactor the code and can make it better/generic. MTLS as a type of authentication is a slight change to TLS part (where we configure `ServerConnectors` to accept client certs). Even I want to push as much code upstream as possible :)